akka-distributed-data/src/main/scala/akka/cluster/ddata/DurableStore.scala

```diff
@@ -0,0 +1,245 @@
+/**
+ * Copyright (C) 2016 Lightbend Inc. <http://www.lightbend.com>
+ */
+package akka.cluster.ddata
+
+import java.io.File
+import java.util.concurrent.TimeUnit
+
+import scala.collection.JavaConverters._
+import scala.collection.breakOut
+import scala.concurrent.duration._
+import scala.util.control.NonFatal
+
+import akka.actor.Actor
+import akka.actor.ActorLogging
+import akka.actor.ActorRef
+import akka.actor.ActorSystem
+import akka.actor.DeadLetterSuppression
+import akka.actor.ExtendedActorSystem
+import akka.actor.Props
+import akka.cluster.Cluster
+import akka.cluster.ddata.Replicator.Internal.DataEnvelope
+import akka.serialization.Serialization
+import akka.serialization.SerializationExtension
+import akka.serialization.SerializerWithStringManifest
+import com.typesafe.config.Config
+import org.mapdb.DBMaker
+import org.mapdb.DataIO
+import org.mapdb.DataInput2
+import org.mapdb.DataOutput2
+import org.mapdb.HTreeMap
+import org.mapdb.Serializer
+
+/**
+ * An actor implementing the durable store for the Distributed Data `Replicator`
+ * have to implement the protocol with the messages defined here.
+ *
+ * At startup the `Replicator` creates the durable store actor and sends the
+ * `Load` message to it. It must then reply with 0 or more `LoadData` messages
+ * followed by one `LoadCompleted` message to the `sender` (the `Replicator`).
+ *
+ * When the `Replicator` needs to store a value it sends a `Store` message
+ * to the durable store actor, which must then reply with the `successMsg` or
+ * `failureMsg` to the `replyTo`.
+ */
+object DurableStore {
+  final case class Store(key: String, data: ReplicatedData, reply: Option[StoreReply])
+  final case class StoreReply(successMsg: Any, failureMsg: Any, replyTo: ActorRef)
+
+  case object Load
+  final case class LoadData(data: Map[String, ReplicatedData])
+  case object LoadCompleted
+}
+
+object MapDbDurableStore {
+  def props(config: Config): Props =
+    Props(new MapDbDurableStore(config))
+
+  private case object Commit extends DeadLetterSuppression
+}
+
+final class MapDbDurableStore(config: Config) extends Actor with ActorLogging {
+  import DurableStore._
+  import MapDbDurableStore.Commit
+
+  val serializer = SerializationExtension(context.system).serializerFor(classOf[DataEnvelope])
+  val commitInterval = config.getString("mapdb.commit-interval").toLowerCase match {
+    case "off" ⇒ Duration.Zero
+    case _     ⇒ config.getDuration("mapdb.commit-interval", MILLISECONDS).millis
+  }
+
+  val db = {
+    val file = config.getString("mapdb.file") match {
+      case path if path.endsWith(".ddata") ⇒
+        new File(path)
+      case dir ⇒
+        new File(dir, s"${context.system.name}-${self.path.parent.name}-${Cluster(context.system).selfAddress.port.get}.ddata")
+    }
+    file.getParentFile.mkdirs()
```

Comment:
Yeah permissions have better apis with Paths nowadays. Let's take it to a separate discussion though. In general about all our files then